packages/firestore/src/local/tab_notification_channel.ts

```diff
@@ -0,0 +1,393 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Code, FirestoreError } from '../util/error';
+import { BatchId, InstanceKey, TargetId } from '../core/types';
+import { assert } from '../util/assert';
+import { debug, error } from '../util/log';
+import { primitiveComparator } from '../util/misc';
+import { SortedSet } from '../util/sorted_set';
+import {
+  validateNamedArray,
+  validateNamedOptionalType,
+  validateNamedType
+} from '../util/input_validation';
+
+// Prefix keys used in WebStorage.
+const FIRESTORE_PREFIX = 'fs';
+const INSTANCE_ROW_PREFIX = 'instances';
+
+const LOG_TAG = 'TabNotificationChannel';
+
+/**
+ * WebStorage of the Firestore client.
+ *
+ * Firestore uses WebStorage for cross-tab notifications to propagate the
+ * state of pending mutation batches and query targets. All Firestore tabs are
+ * responsible for persisting their instance state as their in-memory state
+ * changes. Primary tabs will further update the state of each mutation batch
+ * and query target.
+ */
+export interface TabNotificationChannel {
```

Comment:
`TabNotificationChannel` is a name that Gil suggested in the earlier PR. My original name for this was just `WebStorage`, which is admittedly strictly worse. 

I actually prefer the term 'instance'. In general, I have been using it in this code, partly because names like `instance_key` read better, and partly because the implementation doesn't really have anything tab specific. It's about instance communication, and could work just as well on desktop for C++. I think a good name for the data in this class would be `InstanceMetadata`.

So, that brings us back to naming. Here are some more suggestions:
- InstanceMetadataNotifier
- InstanceMetadataChannel
- InstanceMetadataBroadcast
- InstanceMetadataRelay
- SynchronizedInstanceState

For now, I picked InstanceMetdataChannel. Despite your argument against it, it reads a little less clunky to me.